src/qat/compiler/validation_passes.py

```diff
@@ -0,0 +1,145 @@
+                f"Expected to find an instance of {QuantumExecutionEngine} in arguments list, but got {engine} instead"
+            )
+
+        instruction_length = len(builder.instructions)
+        if instruction_length > engine.max_instruction_len:
+            raise ValueError(
+                f"Program too large to be run in a single block on current hardware. "
+                f"{instruction_length} instructions."
+            )
+
+        for inst in builder.instructions:
+            if isinstance(inst, Acquire) and not inst.channel.acquire_allowed:
+                raise ValueError(
+                    f"Cannot perform an acquire on the physical channel with id "
+                    f"{inst.channel.physical_channel}"
+                )
+            if isinstance(inst, (Pulse, CustomPulse)):
+                duration = inst.duration
+                if isinstance(duration, Number) and duration > MaxPulseLength:
+                    if (
+                        not qatconfig.DISABLE_PULSE_DURATION_LIMITS
+                    ):  # Do not throw error if we specifically disabled the limit checks.
+                        # TODO: Add a lower bound for the pulse duration limits as well in a later PR,
+                        # which is specific to each hardware model and can be stored as a member variables there.
+                        raise ValueError(
+                            f"Max Waveform width is {MaxPulseLength} s "
+                            f"given: {inst.duration} s"
+                        )
+                elif isinstance(duration, Variable):
+                    values = next(
+                        iter(
+                            [
+                                sw.variables[duration.name]
+                                for sw in builder.instructions
+                                if isinstance(sw, Sweep)
+                                and duration.name in sw.variables.keys()
+                            ]
+                        )
+                    )
+                    if np.max(values) > MaxPulseLength:
+                        if not qatconfig.DISABLE_PULSE_DURATION_LIMITS:
+                            raise ValueError(
+                                f"Max Waveform width is {MaxPulseLength} s "
+                                f"given: {values} s"
+                            )
+
+
+class ReadoutValidation(ValidationPass):
+    """
+    Extracted from LiveDeviceEngine.optimize()
+    """
+
+    def run(self, builder: InstructionBuilder, res_mgr: ResultManager, *args, **kwargs):
+        model = next((a for a in args if isinstance(a, QuantumHardwareModel)), None)
+
+        if not model:
+            model = kwargs.get("model", None)
+
+        if not model or not isinstance(model, QuantumHardwareModel):
+            raise ValueError(
+                f"Expected to find an instance of {QuantumHardwareModel} in arguments list, but got {model} instead"
+            )
```

Comment:
Is there a reason we don't explicitly add model as an argument to run?

`run(self, builder: InstructionBuilder, res_mgr: ResultManager, model : QuantumHardwareModel, *args, **kwargs)`

Reply:
mainly because the model is not finalised yet. we are refactoring it and I am reluctant on making assumptions early on

Reply:
Yes, I thought that's what you would say and on reflection I agree.